builtin/credential/okta/backend.go

```diff
@@ -106,6 +119,71 @@ func (b *backend) Login(ctx context.Context, req *logical.Request, username stri
 		}
 		return nil, logical.ErrorResponse("okta authentication failed"), nil, nil
 
+	case "MFA_ENROLL", "MFA_ENROLL_ACTIVATE":
+		if b.Logger().IsDebug() {
+			b.Logger().Debug("auth/okta: user must enroll or complete mfa enrollment", "user", username)
+		}
+		return nil, logical.ErrorResponse("okta authentication failed: you must complete MFA enrollment to continue"), nil, nil
+
+	case "MFA_REQUIRED":
+		factorAvailable := false
+
+		var selectedFactor mfaFactor
+		// only okta push is currently supported
+		for _, v := range result.Embedded.Factors {
+			if v.Type == "push" {
+				factorAvailable = true
+				selectedFactor = v
+			}
+		}
+
+		if factorAvailable {
+			requestPath := fmt.Sprintf("authn/factors/%s/verify", selectedFactor.Id)
+			payload := map[string]interface{}{
+				"stateToken": result.StateToken,
+			}
+			verifyReq, err := client.NewRequest("POST", requestPath, payload)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+
+			rsp, err := client.Do(verifyReq, &result)
+			if err != nil {
+				return nil, logical.ErrorResponse(fmt.Sprintf("Okta auth failed: %v", err)), nil, nil
+			}
+			if rsp == nil {
+				return nil, logical.ErrorResponse("okta auth backend unexpected failure"), nil, nil
+			}
+			for result.Status == "MFA_CHALLENGE" {
+				switch result.FactorResult {
+				case "WAITING":
+					verifyReq, err := client.NewRequest("POST", requestPath, payload)
+					rsp, err := client.Do(verifyReq, &result)
+					if err != nil {
+						return nil, logical.ErrorResponse(fmt.Sprintf("Okta auth failed checking loop: %v", err)), nil, nil
+					}
+					if rsp == nil {
+						return nil, logical.ErrorResponse("okta auth backend unexpected failure"), nil, nil
+					}
+					// okta has a documented rate limit of ~1tps, so we try to respect that here
+					time.Sleep(500 * time.Millisecond)
```

Comment:
Because this can end up taking a very long time, wrap this sleep in a select case with ctx.Done() and time.After() so if Vault needs to shut down it's not held hostage by outstanding requests.